## Упражнения по библиотеке Numpy

In [1]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [2]:
m=np.arange(42)
m[(3<m) & (m<8)]*=-1
print(m)

[ 0  1  2  3 -4 -5 -6 -7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41]


**2.** Заменить максимальный элемент случайного массива на 0

In [3]:
arr=np.random.random(10)
arr[arr.argmax()]=0
arr

array([0.45987062, 0.58590657, 0.        , 0.63558934, 0.47271477,
       0.70484741, 0.8908968 , 0.37845539, 0.57619866, 0.41578434])

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [4]:
arrays = ([1, 2, 3], [4, 5], [6, 7])
arrays = [np.asarray(a) for a in arrays]
shape = map(len, arrays)

ix = np.indices(shape)
ix = ix.reshape(len(arrays), -1).T

for n, arr in enumerate(arrays):
    ix[:, n] = arrays[n][ix[:, n]]

print(ix)

[[1 4 6]
 [1 4 7]
 [1 5 6]
 [1 5 7]
 [2 4 6]
 [2 4 7]
 [2 5 6]
 [2 5 7]
 [3 4 6]
 [3 4 7]
 [3 5 6]
 [3 5 7]]


**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [5]:
# your code
A = np.random.randint(0,5,(8,3))
B = np.random.randint(0,5,(2,2))
print(A)
print()
print(B)
print()

C = (A[..., np.newaxis, np.newaxis] == B)
rows = (C.sum(axis=(1,2,3)) >= B.shape[1]).nonzero()[0]
print(rows)

[[2 0 1]
 [0 4 4]
 [1 1 0]
 [2 1 3]
 [2 3 2]
 [0 0 0]
 [1 0 4]
 [0 3 0]]

[[4 4]
 [2 0]]

[0 1 4 5 6 7]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [6]:
# your code
Z = np.random.randint(8,10,(10,3))
print(Z)
print()

E = np.logical_and.reduce(Z[:,1:] == Z[:,:-1], axis=1)
U = Z[~E]
print(U)

[[9 8 8]
 [9 8 9]
 [9 8 8]
 [9 9 8]
 [9 8 9]
 [9 9 8]
 [9 9 8]
 [9 9 9]
 [8 8 8]
 [9 9 8]]

[[9 8 8]
 [9 8 9]
 [9 8 8]
 [9 9 8]
 [9 8 9]
 [9 9 8]
 [9 9 8]
 [9 9 8]]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [7]:
# your code
a = [[1, 0, 1, 0, 0, 1],
     [1, 1, 1, 0, 0, 1],
     [1, 0, 1, 0, 0, 1],
     [1, 0, 1, 0, 1, 1],
     [1, 0, 1, 0, 0, 1]]

b = np.array(a)

b=np.unique(b, axis=0)

b

array([[1, 0, 1, 0, 0, 1],
       [1, 0, 1, 0, 1, 1],
       [1, 1, 1, 0, 0, 1]])

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [8]:
#without numpy
X = [[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]
mult=1
m = min(len(X), len(X[0]))
for i in range(m):
    if X[i][i] !=0:
        mult*=X[i][i]
    i+=1
print(mult)

#numpy
Y = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])
print(np.prod(np.diag(Y)[np.diag(Y)!=0]))

3
3


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [9]:
x = [1, 2, 2, 4]
y = [4, 2, 1, 2]

# without numpy
print(sorted(x)==sorted(y))

#from collections import Counter
#print(Counter(x)==Counter(y))

# numpy
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])
def are_multisets_equal_vect(x, y):
    x_nums, x_counts = np.unique(x, return_counts=True)
    y_nums, y_counts = np.unique(y, return_counts=True)
    if(np.shape(x_nums) != np.shape(y_nums)):
        return False
    if(np.any(x_nums != y_nums) or np.any(x_counts != y_counts)):
        return False
    return True
print(are_multisets_equal_vect(x,y))

True
True


* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. 
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [10]:
# your code

#true python
x = [6, 2, 0, 3, 0, 0, 5, 7, 0]
print(max(x[i] for i in range(1, len(x)) if not x[i-1]))

#numpy aka np
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])
zero = x==0
print(x[1:][zero[:-1]].max())

5
5


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [11]:
# your code

#numpy 
x = np.array([2, 2, 2, 3, 3, 3, 5])
a=np.unique(x, return_counts=True)
print(a[0],a[1])
#python
x = [2, 2, 2, 3, 3, 3, 5]


[2 3 5] [3 3 1]


* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [21]:
# your code
#import scipy
from scipy.spatial import distance
a=np.array([[0,1],[1,2],[2,3],[3,4],[4,5]])
b=np.array([[5,6],[6,7],[7,8],[8,9],[9,0]])

print(scipy.spatial.distance.cdist(a,b))

print()

#numpy
A=np.array([[0,1],[1,2],[2,3],[3,4],[4,5]])
B=np.array([[5,6],[6,7],[7,8],[8,9],[9,0]])

matrix=np.zeros((len(A), len(B)))
i=0
for a in A:
    j=0
    for b in B:
        matrix[i,j]=np.linalg.norm(a-b)
        j+=1
    i+=1
print(matrix)
print()

#true python
A=[[0,1],[1,2],[2,3],[3,4],[4,5]]
B=[[5,6],[6,7],[7,8],[8,9],[9,0]]

matrix = [0]*len(A)
for i in range(len(A)):
    matrix[i]=[0]*len(B)
i=0
for x1,y1 in A:
    j=0
    for x2,y2 in B:
        matrix[i][j]=((x1-x2)**2+(y1-y2)**2)**0.5
        j+=1
    i+=1
print(matrix)


#timeit
import timeit
a=np.array([[0,1],[1,2],[2,3],[3,4],[4,5]])
b=np.array([[5,6],[6,7],[7,8],[8,9],[9,0]])
#timeit.timeit('scipy.spatial.distance.cdist(a,b)')

[[ 7.07106781  8.48528137  9.89949494 11.3137085   9.05538514]
 [ 5.65685425  7.07106781  8.48528137  9.89949494  8.24621125]
 [ 4.24264069  5.65685425  7.07106781  8.48528137  7.61577311]
 [ 2.82842712  4.24264069  5.65685425  7.07106781  7.21110255]
 [ 1.41421356  2.82842712  4.24264069  5.65685425  7.07106781]]

[[ 7.07106781  8.48528137  9.89949494 11.3137085   9.05538514]
 [ 5.65685425  7.07106781  8.48528137  9.89949494  8.24621125]
 [ 4.24264069  5.65685425  7.07106781  8.48528137  7.61577311]
 [ 2.82842712  4.24264069  5.65685425  7.07106781  7.21110255]
 [ 1.41421356  2.82842712  4.24264069  5.65685425  7.07106781]]

[[7.0710678118654755, 8.48528137423857, 9.899494936611665, 11.313708498984761, 9.055385138137417], [5.656854249492381, 7.0710678118654755, 8.48528137423857, 9.899494936611665, 8.246211251235321], [4.242640687119285, 5.656854249492381, 7.0710678118654755, 8.48528137423857, 7.615773105863909], [2.8284271247461903, 4.242640687119285, 5.656854249492381, 7.071067811865

_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [ ]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [13]:
calorie_stats = np.loadtxt("data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [14]:
# your code
average_calories=np.average(calorie_stats)
average_calories

106.88311688311688

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [15]:
# your code
calorie_stats_sorted=np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [16]:
# your code
median_calories=np.median(calorie_stats_sorted)
median_calories

110.0

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [17]:
# your code
#print(np.percentile(calorie_stats_sorted, np.arange(len(calorie_stats_sorted))))

for i in np.arange(len(calorie_stats_sorted)):
    if np.percentile(calorie_stats_sorted, i) <= 60 :
        print(np.percentile(calorie_stats_sorted, i))
    else :
        nth_percentile = np.percentile(calorie_stats_sorted, i)
        break
print(nth_percentile)
    

50.0
50.0
50.0
55.599999999999994
70.0


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [18]:
# your code
more_calories=(len(calorie_stats_sorted)-i+1)/len(calorie_stats_sorted)*100
print(more_calories)

96.1038961038961


7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [19]:
# your code
calorie_std=np.std(calorie_stats_sorted)
print(calorie_std)

19.35718533390827


8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

# Вывод:
## посчитали, что калорийность нашей продукции ниже, чем у 96% конкурентов это отличный аргумент в выборе злаков нашей компании
##### было бы неплохо еще показать насколько процентов в среднем ниже калорийность
## также высчитали среднее значение и медиану и сравнили их(сильных различий нет), посчитали отклонение
##### если высчитать коэффициент вариации, то можно будет судить о том как значения из выборки отличаются от среднего значения